File: buildSrc/src/main/java/wolkenschloss/download/BaseImageExtension.java
package wolkenschloss.download;

import org.gradle.api.Project;
import org.gradle.api.file.DirectoryProperty;
import org.gradle.api.file.RegularFileProperty;
import org.gradle.api.model.ObjectFactory;
import org.gradle.api.provider.Property;

import javax.inject.Inject;
import wolkenschloss.Directories;

public abstract class BaseImageExtension {

    public static final String DEFAULT_DOWNLOAD_URL = "https://cloud-images.ubuntu.com/focal/current/focal-server-cloudimg-amd64-disk-kvm.img";
    public static final String DEFAULT_DISTRIBUTION_NAME = "ubuntu-20.04";
    public static final String APP_NAME = "testbed";

    @Inject
    public abstract ObjectFactory getObjects();

    public abstract Property<String> getUrl();

    public abstract Property<String> getName();

    public abstract DirectoryProperty getDownloadDir();

    public abstract DirectoryProperty getDistributionDir();

    public abstract RegularFileProperty getBaseImageFile();

    public void initialize(Project project) {
        getUrl().convention(DEFAULT_DOWNLOAD_URL);
        getName().convention(DEFAULT_DISTRIBUTION_NAME);

        getDownloadDir().set(project.getLayout()
                .getProjectDirectory()
                .dir(Directories.getTestbedHome().toFile().getAbsolutePath()));

        getDistributionDir().convention(getDownloadDir().dir(getName()));

        var parts = getUrl().get().split("/");
        var basename = parts[parts.length - 1];

        getBaseImageFile().convention(getDistributionDir().file(basename));
    }
}

File: buildSrc/src/main/kotlin/wolkenschloss/gradle/testbed/download/BaseImageExtension.kt
package wolkenschloss.gradle.testbed.download

import org.gradle.api.Project
import org.gradle.api.file.DirectoryProperty
import org.gradle.api.file.RegularFileProperty
import org.gradle.api.model.ObjectFactory
import org.gradle.api.provider.Property
// TODO: In der Hierarchy nicht nach oben greifen
import wolkenschloss.gradle.testbed.Directories
import javax.inject.Inject

abstract class BaseImageExtension {
    @get:Inject
    abstract val objects: ObjectFactory?
    abstract val url: Property<String>
    abstract val name: Property<String?>
    abstract val downloadDir: DirectoryProperty
    abstract val distributionDir: DirectoryProperty
    abstract val baseImageFile: RegularFileProperty
    fun initialize(project: Project) {
        url.convention(DEFAULT_DOWNLOAD_URL)
        name.convention(DEFAULT_DISTRIBUTION_NAME)
        downloadDir.set(
            project.layout
                .projectDirectory
                .dir(Directories.testbedHome.toFile().absolutePath)
        )
        distributionDir.convention(downloadDir.dir(name))
        val parts = url.get().split("/".toRegex()).toTypedArray()
        val basename = parts[parts.size - 1]
        baseImageFile.convention(distributionDir.file(basename))
    }

    companion object {
        const val DEFAULT_DOWNLOAD_URL =
            "https://cloud-images.ubuntu.com/focal/current/focal-server-cloudimg-amd64-disk-kvm.img"
        const val DEFAULT_DISTRIBUTION_NAME = "ubuntu-20.04"
        const val APP_NAME = "testbed"
    }
}